File: c/src/health.c
#include "../include/af_health.h"
#include "../include/af_broker.h"
#include "../include/af_types.h"

#include <time.h>
#include <stdio.h>
#include <string.h>

void af_health_init(af_health_monitor_t *hm)
{
    hm->start_time = (int64_t)time(NULL);
}

af_health_snapshot_t af_health_check(const af_health_monitor_t *hm,
                                      const af_broker_t *broker,
                                      double equity,
                                      double daily_pnl_pct,
                                      int guard_state,
                                      int signals_processed,
                                      int signals_rejected)
{
    af_health_snapshot_t snap;
    memset(&snap, 0, sizeof(snap));

    int64_t now = (int64_t)time(NULL);
    snap.uptime_sec       = now - hm->start_time;
    snap.broker_connected = broker ? broker->is_connected(broker) : 0;
    snap.equity           = equity;
    snap.daily_pnl_pct    = daily_pnl_pct;
    snap.guard_state      = guard_state;
    snap.signals_processed = signals_processed;
    snap.signals_rejected  = signals_rejected;

    /* Count open positions */
    if (broker && snap.broker_connected) {
        snap.open_positions = broker->get_positions(broker, NULL, 0);
    } else {
        snap.open_positions = 0;
    }

    /* Determine status string
     * guard_state: 0=GREEN, 1=YELLOW, 2=RED */
    if (guard_state == 2) {
        strncpy(snap.status, "CRITICAL", sizeof(snap.status) - 1);
    } else if (guard_state == 1 || daily_pnl_pct <= -5.0) {
        strncpy(snap.status, "WARN", sizeof(snap.status) - 1);
    } else {
        strncpy(snap.status, "OK", sizeof(snap.status) - 1);
    }
    snap.status[sizeof(snap.status) - 1] = '\0';

    return snap;
}

void af_health_print_json(const af_health_snapshot_t *snap)
{
    printf("{"
           "\"uptime_sec\":%lld,"
           "\"broker_connected\":%d,"
           "\"equity\":%.2f,"
           "\"daily_pnl_pct\":%.4f,"
           "\"guard_state\":%d,"
           "\"open_positions\":%d,"
           "\"signals_processed\":%d,"
           "\"signals_rejected\":%d,"
           "\"status\":\"%s\""
           "}\n",
           (long long)snap->uptime_sec,
           snap->broker_connected,
           snap->equity,
           snap->daily_pnl_pct,
           snap->guard_state,
           snap->open_positions,
           snap->signals_processed,
           snap->signals_rejected,
           snap->status);
    fflush(stdout);
}

File: js/src/health.js
/**
 * Health-check snapshot and monitor — Round 11 (24/7 Ops).
 * Mirrors the Python HealthMonitor / HealthSnapshot contract.
 */

import { GuardState } from './risk.js';

// ── HealthSnapshot ────────────────────────────────────────────────────────────

export class HealthSnapshot {
  /**
   * @param {object} opts
   * @param {number}  opts.uptimeSec
   * @param {boolean} opts.brokerConnected
   * @param {number}  opts.equity
   * @param {number}  opts.dailyPnlPct
   * @param {string|number} opts.guardState   GuardState value (or numeric sentinel)
   * @param {number}  opts.openPositions
   * @param {number}  opts.signalsProcessed
   * @param {number}  opts.signalsRejected
   * @param {string}  opts.status            'OK' | 'WARN' | 'CRITICAL'
   */
  constructor({
    uptimeSec, brokerConnected, equity, dailyPnlPct, guardState,
    openPositions, signalsProcessed, signalsRejected, status,
  }) {
    this.uptimeSec        = uptimeSec;
    this.brokerConnected  = brokerConnected;
    this.equity           = equity;
    this.dailyPnlPct      = dailyPnlPct;
    this.guardState       = guardState;
    this.openPositions    = openPositions;
    this.signalsProcessed = signalsProcessed;
    this.signalsRejected  = signalsRejected;
    this.status           = status;
  }

  /** Plain object representation — suitable for structured logging. */
  toDict() { return { ...this }; }

  /** Compact JSON string. */
  toJSON() { return JSON.stringify(this.toDict()); }
}

// ── HealthMonitor ─────────────────────────────────────────────────────────────

export class HealthMonitor {
  constructor() {
    this._startTime = Date.now();
  }

  /** Seconds since this monitor was constructed. */
  get uptimeSec() { return Math.floor((Date.now() - this._startTime) / 1000); }

  /**
   * Build a health snapshot from the current runtime state.
   *
   * guardState accepts:
   *   - GuardState string values ('green' | 'yellow' | 'red')
   *   - Numeric sentinels used by tests/legacy code (0 = GREEN, 1 = YELLOW, 2 = RED)
   *
   * @param {object|null} broker           IBroker-compatible instance (may be null)
   * @param {number}      equity           Current account equity
   * @param {number}      dailyPnlPct      Daily P&L as a fraction (e.g. -0.03 = -3%)
   * @param {string|number} guardState     GuardState value (or numeric sentinel)
   * @param {number}      [signalsProcessed=0]
   * @param {number}      [signalsRejected=0]
   * @returns {HealthSnapshot}
   */
  check(broker, equity, dailyPnlPct, guardState, signalsProcessed = 0, signalsRejected = 0) {
    const uptimeSec      = this.uptimeSec;
    const brokerConnected = broker ? broker.isConnected() : false;

    let openPositions = 0;
    if (broker && brokerConnected) {
      try { openPositions = broker.getPositions().length; } catch { /* ignore */ }
    }

    // Derive status from guardState (string or numeric) and dailyPnlPct.
    // Numeric sentinels: 0 = GREEN, 1 = YELLOW, 2 = RED  (matches GuardState ordering).
    const isRed =
      guardState === GuardState.RED    ||
      guardState === 'red'             ||
      guardState === 2;

    const isYellow =
      guardState === GuardState.YELLOW ||
      guardState === 'yellow'          ||
      guardState === 1;

    let status;
    if (isRed) {
      status = 'CRITICAL';
    } else if (isYellow || dailyPnlPct < -0.03) {
      status = 'WARN';
    } else {
      status = 'OK';
    }

    return new HealthSnapshot({
      uptimeSec,
      brokerConnected,
      equity,
      dailyPnlPct,
      guardState,
      openPositions,
      signalsProcessed,
      signalsRejected,
      status,
    });
  }
}
